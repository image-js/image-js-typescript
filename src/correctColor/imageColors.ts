export const polishImageData = {
  colors: [
    [41, 146, 219],
    [223, 42, 41],
    [244, 164, 59],
    [90, 57, 48],
    [133, 153, 202],
    [205, 34, 119],
    [80, 157, 61],
    [84, 41, 95],
    [94, 108, 53],
    [210, 63, 77],
    [243, 230, 71],
    [22, 52, 167],
    [104, 132, 168],
    [250, 119, 47],
    [244, 244, 245],
    [29, 31, 34],
    [96, 99, 99],
    [125, 129, 131],
    [165, 168, 165],
    [201, 202, 193],
  ],
  type: 'v1.0',
};
