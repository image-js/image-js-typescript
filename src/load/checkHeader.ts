/**
 * Check the presence of a header in a byte array.
 * @private
 * @param arr - The array to check.
 * @param header - The header to check the presence of.
 */
export default function checkHeader(
  arr: Uint8Array,
  header: number[]
): boolean {
  if (arr.length < header.length) {
    return false;
  }
  for (let i = 0; i < header.length; i++) {
    if (header[i] !== arr[i]) {
      return false;
    }
  }
  return true;
}
