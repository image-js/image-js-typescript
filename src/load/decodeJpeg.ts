import { decode } from 'jpeg-js';

import { Image, ImageKind } from '../Image';

/**
 * Decode a jpeg. See the jpeg-js npm module.
 * @param buffer The data to decode
 */
export function decodeJpeg(buffer: Uint8Array): Image {
  const jpeg = decode(buffer, true);
  return new Image(jpeg.width, jpeg.height, {
    data: jpeg.data,
    kind: ImageKind.RGBA
  });
}
