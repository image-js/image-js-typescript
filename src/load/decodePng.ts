import { decode, IDecodedPNG } from 'fast-png';

import { Image, ImageKind, ColorDepth } from '../Image';

/**
 * Decode a png. See the fast-png npm module.
 * @param buffer The data to decode
 */
export function decodePng(buffer: Uint8Array): Image {
  const png = decode(buffer);

  let kind: ImageKind;
  const depth: ColorDepth =
    png.depth === 16 ? ColorDepth.UINT16 : ColorDepth.UINT8;

  if (png.palette) {
    return loadPalettePNG(png);
  }

  switch (png.channels) {
    case 1:
      kind = ImageKind.GREY;
      break;
    case 2:
      kind = ImageKind.GREYA;
      break;
    case 3:
      kind = ImageKind.RGB;
      break;
    case 4:
      kind = ImageKind.RGBA;
      break;
    default:
      throw new Error(`Unexpected number of channels: ${png.channels}`);
  }
  return new Image({
    width: png.width,
    height: png.height,
    kind,
    depth,
    data: png.data
  });
}

function loadPalettePNG(png: IDecodedPNG): Image {
  if (!png.palette) {
    throw new Error(
      'unexpected: there should be a palette when colourType is 3'
    );
  }
  const pixels = png.width * png.height;
  const data = new Uint8Array(pixels * 3);
  const pixelsPerByte = 8 / png.depth;
  const factor = png.depth < 8 ? pixelsPerByte : 1;
  const mask = parseInt('1'.repeat(png.depth), 2);
  let dataIndex = 0;

  for (let i = 0; i < pixels; i++) {
    const index = Math.floor(i / factor);
    let value = png.data[index];
    if (png.depth < 8) {
      value =
        (value >>> (png.depth * (pixelsPerByte - 1 - (i % pixelsPerByte)))) &
        mask;
    }
    const paletteValue = png.palette[value];
    data[dataIndex++] = paletteValue[0];
    data[dataIndex++] = paletteValue[1];
    data[dataIndex++] = paletteValue[2];
  }

  return new Image(png.width, png.height, {
    data
  });
}
