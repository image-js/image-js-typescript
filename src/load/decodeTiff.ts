import { decode } from 'tiff';

import { ImageColorModel } from '..';
import { Image } from '../Image';

type TiffIfd = ReturnType<typeof decode>[number];

/**
 * Decode a TIFF. See the tiff module.
 *
 * @param buffer - The data to decode.
 * @returns The decoded image.
 */
export function decodeTiff(buffer: Uint8Array): Image {
  let result = decode(buffer);
  return getImageFromIFD(result[0]);
  // TODO: handle stacks (many IFDs)
}

/**
 * Create image from a single IFD.
 *
 * @param ifd - The IFD.
 * @returns The decoded image.
 */
function getImageFromIFD(ifd: TiffIfd): Image {
  if (ifd.type === 3) {
    // Palette
    const data = new Uint16Array(3 * ifd.width * ifd.height);
    const palette = ifd.palette as [number, number, number][];
    let ptr = 0;
    for (let index of ifd.data) {
      const color = palette[index];
      data[ptr++] = color[0];
      data[ptr++] = color[1];
      data[ptr++] = color[2];
    }
    return new Image(ifd.width, ifd.height, {
      data,
      // TODO: handle alpha properly
      colorModel: ifd.alpha ? ImageColorModel.RGBA : ImageColorModel.RGB,
      // TODO: handle other bit depths
      depth: 16,
      // TODO: implement metadata
      //meta: getMetadata(ifd),
    });
  } else {
    return new Image(ifd.width, ifd.height, {
      // TODO: handle float data
      // @ts-expect-error float data not handled yet
      data: ifd.data,
      depth: ifd.bitsPerSample,
      colorModel:
        ifd.type === 2
          ? ifd.alpha
            ? ImageColorModel.RGBA
            : ImageColorModel.RGB
          : ifd.alpha
          ? ImageColorModel.GREYA
          : ImageColorModel.GREY,
      // meta: getMetadata(ifd),
    });
  }
}
