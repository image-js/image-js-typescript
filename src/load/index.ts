export * from './decode';
export * from './decodeJpeg';
export * from './decodePng';
export * from './read';
export * from './readCanvas';
export * from './readImg';
