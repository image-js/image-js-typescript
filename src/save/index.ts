export * from './encode';
export * from './encodePng';
export * from './encodeJpeg';
export * from './write';
export * from './writeCanvas';
