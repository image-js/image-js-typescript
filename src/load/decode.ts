import { Image } from '../Image';

import checkHeader from './checkHeader';
import { decodePng } from './decodePng';
import { decodeJpeg } from './decodeJpeg';

const pngHeader = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];
const jpegHeader = [0xff, 0xd8, 0xff];

/**
 * Decode input data. Data format is automatically detected.
 * Possible formats: png and jpeg.
 * @param data - data to decode
 */
export function decode(data: ArrayBufferView): Image {
  const typedArray = new Uint8Array(
    data.buffer,
    data.byteOffset,
    data.byteLength
  );
  if (checkHeader(typedArray, pngHeader)) {
    return decodePng(typedArray);
  } else if (checkHeader(typedArray, jpegHeader)) {
    return decodeJpeg(typedArray);
  } else {
    throw new Error('unrecognized data format');
  }
}
