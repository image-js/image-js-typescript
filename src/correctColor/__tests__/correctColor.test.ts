import { readSync } from '../../load';
import { writeSync } from '../../save';
import {
  arraysToRgbColors,
  correctColor,
  formatReferenceForMlr,
} from '../correctColor';
import { polishImageData } from '../imageColors';
import { referenceQpCard } from '../referenceQpCard';

describe('formatReferenceForMlr', () => {
  it('convert reference QP card', () => {
    const result = formatReferenceForMlr(referenceQpCard);
    expect(result).toMatchSnapshot();
    expect(result.r).toHaveLength(20);
    expect(result.g).toHaveLength(20);
    expect(result.b).toHaveLength(20);
  });
});

describe('correctColor', () => {
  it('small test image', () => {
    const image = testUtils.load('opencv/test.png');

    const inputData = arraysToRgbColors(polishImageData.colors);

    const result = correctColor(image, inputData, referenceQpCard);

    writeSync('src/correctColor/__tests__/correctColor-test.png', result);
  });

  it('polish scan data', () => {
    const image = readSync('src/correctColor/__tests__/pol_example.png');

    const inputData = arraysToRgbColors(polishImageData.colors);

    const result = correctColor(image, inputData, referenceQpCard);

    writeSync('src/correctColor/__tests__/correctColor.png', result);
  });
});
