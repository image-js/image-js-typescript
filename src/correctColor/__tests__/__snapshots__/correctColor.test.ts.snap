// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatReferenceForMLR convert reference QP card 1`] = `
Object {
  "b": Array [
    Array [
      210,
    ],
    Array [
      65,
    ],
    Array [
      58,
    ],
    Array [
      60,
    ],
    Array [
      207,
    ],
    Array [
      116,
    ],
    Array [
      54,
    ],
    Array [
      99,
    ],
    Array [
      60,
    ],
    Array [
      92,
    ],
    Array [
      55,
    ],
    Array [
      161,
    ],
    Array [
      172,
    ],
    Array [
      58,
    ],
    Array [
      248,
    ],
    Array [
      38,
    ],
    Array [
      105,
    ],
    Array [
      150,
    ],
    Array [
      177,
    ],
    Array [
      210,
    ],
  ],
  "g": Array [
    Array [
      141,
    ],
    Array [
      59,
    ],
    Array [
      180,
    ],
    Array [
      69,
    ],
    Array [
      161,
    ],
    Array [
      26,
    ],
    Array [
      155,
    ],
    Array [
      51,
    ],
    Array [
      119,
    ],
    Array [
      85,
    ],
    Array [
      232,
    ],
    Array [
      47,
    ],
    Array [
      136,
    ],
    Array [
      142,
    ],
    Array [
      249,
    ],
    Array [
      42,
    ],
    Array [
      104,
    ],
    Array [
      146,
    ],
    Array [
      179,
    ],
    Array [
      216,
    ],
  ],
  "r": Array [
    Array [
      0,
    ],
    Array [
      236,
    ],
    Array [
      254,
    ],
    Array [
      106,
    ],
    Array [
      122,
    ],
    Array [
      224,
    ],
    Array [
      41,
    ],
    Array [
      101,
    ],
    Array [
      110,
    ],
    Array [
      227,
    ],
    Array [
      253,
    ],
    Array [
      0,
    ],
    Array [
      94,
    ],
    Array [
      255,
    ],
    Array [
      246,
    ],
    Array [
      39,
    ],
    Array [
      102,
    ],
    Array [
      144,
    ],
    Array [
      177,
    ],
    Array [
      214,
    ],
  ],
}
`;

exports[`formatReferenceForMlr convert reference QP card 1`] = `
Object {
  "b": Array [
    Array [
      210,
    ],
    Array [
      65,
    ],
    Array [
      58,
    ],
    Array [
      60,
    ],
    Array [
      207,
    ],
    Array [
      116,
    ],
    Array [
      54,
    ],
    Array [
      99,
    ],
    Array [
      60,
    ],
    Array [
      92,
    ],
    Array [
      55,
    ],
    Array [
      161,
    ],
    Array [
      172,
    ],
    Array [
      58,
    ],
    Array [
      248,
    ],
    Array [
      38,
    ],
    Array [
      105,
    ],
    Array [
      150,
    ],
    Array [
      177,
    ],
    Array [
      210,
    ],
  ],
  "g": Array [
    Array [
      141,
    ],
    Array [
      59,
    ],
    Array [
      180,
    ],
    Array [
      69,
    ],
    Array [
      161,
    ],
    Array [
      26,
    ],
    Array [
      155,
    ],
    Array [
      51,
    ],
    Array [
      119,
    ],
    Array [
      85,
    ],
    Array [
      232,
    ],
    Array [
      47,
    ],
    Array [
      136,
    ],
    Array [
      142,
    ],
    Array [
      249,
    ],
    Array [
      42,
    ],
    Array [
      104,
    ],
    Array [
      146,
    ],
    Array [
      179,
    ],
    Array [
      216,
    ],
  ],
  "r": Array [
    Array [
      0,
    ],
    Array [
      236,
    ],
    Array [
      254,
    ],
    Array [
      106,
    ],
    Array [
      122,
    ],
    Array [
      224,
    ],
    Array [
      41,
    ],
    Array [
      101,
    ],
    Array [
      110,
    ],
    Array [
      227,
    ],
    Array [
      253,
    ],
    Array [
      0,
    ],
    Array [
      94,
    ],
    Array [
      255,
    ],
    Array [
      246,
    ],
    Array [
      39,
    ],
    Array [
      102,
    ],
    Array [
      144,
    ],
    Array [
      177,
    ],
    Array [
      214,
    ],
  ],
}
`;
